cmd/object-handlers.go

```diff
@@ -1672,6 +1677,361 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 	})
 }
 
+type tarObjectData struct {
+	Data  []byte
+	Name  string
+	Info  os.FileInfo
+	Error error
+}
+
+func extractTARGZ(r io.Reader) (<-chan tarObjectData, error) {
+	gzf, err := gzip.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+
+	tarReader := tar.NewReader(gzf)
+	chReader := make(chan tarObjectData)
+	go func() {
+		defer close(chReader)
+		bb := bytebufferpool.Get()
+		defer func() {
+			bb.Reset()
+			bytebufferpool.Put(bb)
+		}()
+		for {
+			header, err := tarReader.Next()
+			if err != nil {
+				if err != io.EOF {
+					logger.LogIf(GlobalContext, err)
+					chReader <- tarObjectData{
+						Error: err,
+					}
+				}
+				return
+			}
+
+			name := header.Name
+			if name == slashSeparator {
+				continue
+			}
+
+			switch header.Typeflag {
+			case tar.TypeDir: // = directory
+				chReader <- tarObjectData{
+					Data: nil,
+					Name: trimLeadingSlash(pathJoin(name, slashSeparator)),
+					Info: header.FileInfo(),
+				}
+			case tar.TypeReg: // = regular
```

Comment:
There are a few aliases for files: https://github.com/klauspost/compress/blob/master/s2/cmd/_s2sx/_unpack/main.go#L262